### DISCUSSION TITLE
Tell me how to de-normalize the resource.

### DISCUSSION THREAD

**Discussion by usuall** [zauberzeug/nicegui#1862] [Q&A]
### Question

How do I write a code that returns all resources when the app terminates abnormally, or when it terminates (alt+F4)?

I'm trying to close the sqlite and selenium resources normally and manage them.

# try, catch, finally
Is it possible to implement it by finally putting a code to close the resource?

**Comment by falkoschindler** [comment]
@usuall Sorry, can you give more context or a [minimum reproducible example](https://en.wikipedia.org/wiki/Minimal_reproducible_example) of what you're trying to do? At a first glance this seems to be a general Python question and not specifically related to NiceGUI, in which case StackOverflow would be a much better place to ask.

**Reply by usuall** [reply]
I wrote it below.

**Comment by rodja** [comment]
In general you can use `app.on_shutdown` to register functions which need to be called when the server stops. But these won't be called if the server terminates abnormally. Because... its abnormally.

**Reply by usuall** [reply]
I'm making an app that runs on Windows. Will it be possible?

**Reply by rodja** [reply]
Sure. Have you run into any issues?

**Reply by ed2050** [reply]
You can use python module `atexit` to register a function that's called on _most_ program exits.  It's not invoked in a few cases: fatal errors, uncatchable signals, etc.  If those cases matter to you, serach stackoverflow for ways to deal with them.  In most cases, [atexit](https://docs.python.org/3/library/atexit.html?highlight=atexit) is enough.

**Reply by rodja** [reply]
I suggest you use `app.on_shutdown` for tear down instead of `atexit` because it will be executed on server shutdown, not program exit. The first happens on every server relaod (for example when saving a file). The latter only when the main program exits.
